# Identification data


## Identification data.frame

Let's use the identification from from `msdata`:

```{r mzrvsid, eval = TRUE}
idf <- msdata::ident(full.names = TRUE)
basename(idf)
```

The easiest way to read identification data in `mzIdentML` (often
abbreviated with `mzid`) into R is to read it with `readPSMs()`
function from the [`PSM`](https://rformassspectrometry.github.io/PSM/)
package. The function will parse the file and return a `DataFrame`.

```{r readid, message = FALSE}
library(PSM)
id <- readPSMs(idf)
dim(id)
names(id)
```

```{r, echo = FALSE}
n_matches <- nrow(id)
n_scans <- length(unique(id$spectrumID))
n_seqs <- length(unique(id$sequence))
```

This table contains `r n_matches` matches for `r n_scans` scans and
`r n_seqs` peptides sequences.

```{r}
nrow(id) ## number of matches
length(unique(id$spectrumID)) ## number of scans
length(unique(id$sequence))   ## number of peptide sequences
```

The PSM data are read as is, without and filtering. As we can see
below, we still have all the hits from the forward and reverse (decoy)
databases.

```{r}
table(id$isDecoy)
```

## Keeping all matches

The data contains also contains multiple matches for several
spectra. The table below shows the number of number of spectra that
have 1, 2, ... up to 5 matches.

```{r}
table(table(id$spectrumID))
```

Below, we can see how scan 1774 has 4 matches, all to sequence
`RTRYQAEVR`, which itself matches to 4 different proteins:


```{r}
i <- which(id$spectrumID == "controllerType=0 controllerNumber=1 scan=1774")
id[i, 1:5]
```

If the goal is to keep all the matches, but arranged by scan/spectrum,
one can *reduce* the `DataFrame` object by the `spectrumID` variable,
so that each scan correponds to a single row that still stores all
values[^rownames]:

[^rownames]: The rownames aren't needed here are are removed to reduce
    to output in the the next code chunk display parts of `id2`.

```{r}
id2 <- QFeatures::reduceDataFrame(id, id$spectrumID)
rownames(id2) <- NULL ## rownames not needed here
dim(id2)
```

The resulting object contains a single entrie for scan 1774 with
information for the multiple matches stored as lists within the cells.

```{r}
j <- which(id2$spectrumID == "controllerType=0 controllerNumber=1 scan=1774")
id2[j, ]
```

```{r}
id2[j, "DatabaseAccess"]
```

The is the type of complete identification table that could be used to
annotate an raw mass spectrometry `Spectra` object, as shown below.

## Filtering data

Often, the PSM data is filtered to only retain reliable matches. The
`MSnID` package can be used to set thresholds to attain user-defined
PSM, peptide or protein-level FDRs. Here, we will simply filter out
wrong identification manually.

Here, the `filter()` from the `dplyr` package comes very handy. We
will thus start by convering the `DataFrame` to a `tibble`.

```{r}
library("dplyr")
id_tbl <- tidyr::as_tibble(id)
id_tbl
```

### Exercise {-}

- Remove decoy hits

<details>
```{r, message = FALSE}
id_tbl <- id_tbl %>%
    filter(!isDecoy)
id_tbl
```
</details>

- Keep first rank matches

<details>
```{r}
id_tbl <- id_tbl %>%
    filter(rank == 1)
id_tbl
```
</details>

- Remove non-proteotypic peptides. Start by identifying scans that
  match different proteins. For example scan 4884 matches proteins
  `XXX_ECA3406` and `ECA3415`. Scan 4099 match `XXX_ECA4416_1`,
  `XXX_ECA4416_2` and `XXX_ECA4416_3`. Then remove the scans that
  match any of these proteins.


<details>
```{r}
mltm <-
    id_tbl %>%
    group_by(spectrumID) %>%
    mutate(nProts = length(unique(DatabaseAccess))) %>%
    filter(nProts > 1) %>%
    select(DatabaseAccess, nProts)
mltm
```
```{r}
id_tbl <-
    id_tbl %>%
    filter(!spectrumID %in% mltm$spectrumID)
id_tbl
```
</details>

Which leaves us with `r nrow(id_tbl)` PSMs.

This can also be achieved with the `filterPSMs()` function:

```{r}
id_filtered <- filterPSMs(id)
```

### Exercise {-}

Compare the distribution of raw idenfication scores of the decoy and
non-decoy hits.

```{r}
library(ggplot2)
as_tibble(id) %>%
    ggplot(aes(x = MS.GF.RawScore,
               colour = isDecoy)) +
    geom_density()
```

### Exercise {-}

The `r CRANpkg("tidyverse")`
tools are fit for data wrangling with identification data. Using the
above identification dataframe, calculate the length of each peptide
(you can use `nchar` with the peptide sequence `sequence`) and the
number of peptides for each protein (defined as
`DatabaseDescription`). Plot the length of the proteins against their
respective number of peptides.


<details>
```{r answid1, fig.cap = "Identifcation data wrangling."}
suppressPackageStartupMessages(library("dplyr"))
iddf <- as_tibble(id_filtered) %>%
    mutate(peplen = nchar(sequence))
npeps <- iddf %>%
    group_by(DatabaseAccess) %>%
    tally
iddf <- full_join(iddf, npeps)

library("ggplot2")
ggplot(iddf, aes(x = n, y = DBseqLength)) + geom_point()
```
</details>


## Low level access to id data

There are two packages that can be used to parse `mzIdentML` files,
namely `mzR` (that we have already used for raw data) and `mzID`. The
major difference is that the former leverages C++ code from
`proteowizard` and is hence faster than the latter (which uses the
`XML` R package). They both work in similar ways.

```{r idpkgs, echo = FALSE, eval = FALSE}
x <- RforProteomics:::msDataTab()
sel <- x[, 1] %in% c("Identification")
knitr::kable(x[sel, ])
```

```
|   |Data type      |File format |Data structure |Package |
|:--|:--------------|:-----------|:--------------|:-------|
|4  |Identification |mzIdentML   |mzRident       |mzR     |
|5  |Identification |mzIdentML   |mzID           |mzID    |
```

### `mzID` {-}

The main functions are `mzID` to read the data into a dedicated data
class and `flatten` to transform it into a `data.frame`.

```{r mzid1}
idf
library("mzID")
id <- mzID(idf)
id
```

Various data can be extracted from the `mzID` object, using one the
accessor functions such as `database`, `software`, `scans`, `peptides`,
... The object can also be converted into a `data.frame` using the
`flatten` function.

```{r flatmzid}
head(flatten(id))
```

### `mzR` {-}

The `mzR` interface provides a similar interface. It is however much
faster as it does not read all the data into memory and only extracts
relevant data on demand. It has also accessor functions such as
`softwareInfo`, `mzidInfo`, ... (use `showMethods(classes = "mzRident", where = "package:mzR")`)
to see all available methods.

```{r idmzr}
library("mzR")
id2 <- openIDfile(idf)
id2
softwareInfo(id2)
```

The identification data can be accessed as a `data.frame` with the
`psms` accessor.

```{r psmsmzr}
head(psms(id2))
```

## MS/MS database search

While searches are generally performed using third-party software
independently of R or can be started from R using a `system` call, the
`r Biocpkg("MSGFplus")` package enables to perform a search using the
MSGF+ engine, as illustrated below.


We search the
`TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML.gz`
file against the fasta file from `PXD000001` using `MSGFplus`.

We first download the fasta files from ProteomeXchange:

```{r ex_getfas, eval = FALSE}
fas <- pxget(px, "erwinia_carotovora.fasta")
basename(fas)
```

Below, we setup and run the
search^[In the `runMSGF` call, the memory allocated to the java virtual machine is limited to 1GB. In general, there is no need to specify this argument, unless you experience an error regarding the *maximum heap size*.].

```{r ex_msgfplus, eval = FALSE}
library("MSGFplus")
msgfpar <- msgfPar(database = fas,
                   instrument = 'HighRes',
                   tda = TRUE,
                   enzyme = 'Trypsin',
                   protocol = 'iTRAQ')
idres <- runMSGF(msgfpar, mzf, memory=1000)
idres
```

```
## An mzID object
##
## Software used:   MS-GF+ (version: Beta (v10072))
##
## Rawfile:         /home/lg390/Documents/Teaching/bioc-ms-prot/TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML
##
## Database:        /home/lg390/Documents/Teaching/bioc-ms-prot/erwinia_carotovora.fasta
##
## Number of scans: 5343
## Number of PSM's: 5656
```

A graphical interface to perform the search the data and explore the
results is also available:

```{r msgfgui, eval=FALSE}
library("MSGFgui")
MSGFgui()
```

![The `r Biocpkg("MSGFgui")` interface](./img/MSGFgui.png)

## Adding identification data to raw data

We are goind to use the `sp` object created in the previous chapter
and the `id_filtered` variable generated above.

Identification data (as a `DataFrame`) can be merged into raw data (as
a `Spectra` object) by adding new spectra variables to the appropriate
MS2 spectra. Scans and peptide-spectrum matches can be matched by
their spectrum identifers.

### Exercise {-}

Identify the spectum identifier columns in the `sp` the `id_filtered`
variables.

<details>

In the raw data, it is encoded as `spectrumId`, while in the
identification data, we have `spectrumID`.

```{r}
spectraVariables(sp)
names(id_filtered)
```
</details>

These two data can thus simply be joined using:

```{r}
sp <- joinSpectraData(sp, id_filtered,
                      by.x = "spectrumId",
                      by.y = "spectrumID")
spectraVariables(sp)
```

### Exercise {-}

Verify that the identification data has been added to the correct
spectra.

<details>

1. Let's first verify that no identification data has been added to
   the MS1 scans.

```{r}
all(is.na(filterMsLevel(sp, 1)$sequence))
```

2. They have indeed been added to 56% of the MS2 spectra.

```{r}
sp_2 <- filterMsLevel(sp, 2)
table(is.na(sp_2$sequence))
```

3. Let's compare the precursor/peptide mass to charges

```{r}
sp_2 <- sp_2[!is.na(sp_2$sequence)]
summary(sp_2$precursorMz - sp_2$experimentalMassToCharge)
```
</details>


## Visualising peptide-spectrum matches

Let's choose a MS2 spectrum with a high identication score and plot
it.

```{r}
i <- which(sp$MS.GF.RawScore > 100)[1]
plotSpectra(sp[i])
```

We have seen above that we can add labels to each peak using the
`labels` argument in `plotSpectra()`. The `addFragments()` function
takes a spectrum as input (that is a `Spectra` object of length 1) and
annotates its peaks.

```{r}
addFragments(sp[i])
```

It can be directly used with `plotSpectra()`:

```{r}
plotSpectra(sp[i], labels = addFragments,
            labelPos = 3, labelCol = "steelblue")
```


The annotation of spectra is obtained by simulating fragmentation of a
peptide and matching observed peaks to fragments:

```{r frag}
sp[i]$sequence
calculateFragments(sp[i]$sequence)
```

## Comparing spectra

The `compareSpectra()` can be used to compare spectra (by default,
computing the normalised dot product).

### Exercise {-}

1. Create a new `Spectra` object containing the MS2 spectra with
   sequences `"SQILQQAGTSVLSQANQVPQTVLSLLR"` and
   `"TKGLNVMQNLLTAHPDVQAVFAQNDEMALGALR"`.


<details>
```{r}
k <- which(sp$sequence %in% c("SQILQQAGTSVLSQANQVPQTVLSLLR", "TKGLNVMQNLLTAHPDVQAVFAQNDEMALGALR"))
sp_k <- sp[k]
sp_k
```
</details>

2. Calculate the `r length(sp_k)` by `r length(sp_k)` distance matrix
   between all spectra using `compareSpectra`. See the `?Spectra` man
   page for details. Draw a heatmap of that distance matrix


<details>
```{r}
distmat <- compareSpectra(sp_k)
rownames(distmat) <- colnames(distmat) <- strtrim(sp_k$sequence, 2)
distmat
pheatmap::pheatmap(distmat)
```
</details>

3. Compare the spectra with the plotting function seen previously.

<details>
```{r}
filterIntensity(sp_k, 1e3) %>% plotSpectra(main = sp_k$sequence)
```

```{r}
par(mfrow = c(3, 1))
plotSpectraMirror(sp_k[1], sp_k[2], main = "TK...")
plotSpectraMirror(sp_k[3], sp_k[4], main = "SQ...")
plotSpectraMirror(sp_k[3], sp_k[4], main = "SQ...")
```
</details>


### Exercice {-}

1. Download the 3 first mzML and mzID files from the
   [PXD022816](https://www.ebi.ac.uk/pride/archive/projects/PXD022816)
   project.

2. Generate a `Spectra` object containing only MS2 scans and a table
   of filtered PSMs. Check the quality of the identification data by
   comparing the density of the decoy and forward PSMs id scores for
   each file. What is the proportion if identified MS2 spectra?

3. Join the raw and identification data. Beware though that the
   joining must now be performed by spectrum ids and by files.

4. Extract the PSMs for `"DGSDEPGTAACPNGSFHCTNTGYK"`,
   `"DGQVINETSQHHDDLE"` and `"DSYVGDEAQSK"` and compare and cluster
   the scans.

## Exploration and Assessment of Confidence of LC-MSn Proteomics Identifications using `MSnID`

```{r, echo = FALSE}
unloadNamespace("PSM")
```

The `MSnID` package extracts MS/MS ID data from mzIdentML (leveraging
the `mzID` package) or text files. After collating the search results
from multiple datasets it assesses their identification quality and
optimises filtering criteria to achieve the maximum number of
identifications while not exceeding a specified false discovery
rate. It also contains a number of utilities to explore the MS/MS
results and assess missed and irregular enzymatic cleavages, mass
measurement accuracy, etc.

### Step-by-step work-flow

Let's reproduce parts of the analysis described the `MSnID`
vignette. You can explore more with

```{r, eval = FALSE}
vignette("msnid_vignette", package = "MSnID")
```

The `r Biocpkg("MSnID")` package can be used for post-search filtering
of MS/MS identifications. One starts with the construction of an
`MSnID` object that is populated with identification results that can
be imported from a `data.frame` or from `mzIdenML` files. Here, we
will use the example identification data provided with the package.

```{r idf}
mzids <- system.file("extdata", "c_elegans.mzid.gz", package="MSnID")
basename(mzids)
```

We start by loading the package, initialising the `MSnID` object, and
add the identification result from our `mzid` file (there could of
course be more that one).

```{r msnid1}
library("MSnID")
msnid <- MSnID(".")
msnid <- read_mzIDs(msnid, mzids)
show(msnid)
```

Printing the `MSnID` object returns some basic information such as

* Working directory.
* Number of spectrum files used to generate data.
* Number of peptide-to-spectrum matches and corresponding FDR.
* Number of unique peptide sequences and corresponding FDR.
* Number of unique proteins or amino acid sequence accessions and corresponding FDR.


The package then enables to define, optimise and apply filtering based
for example on missed cleavages, identification scores, precursor mass
errors, etc. and assess PSM, peptide and protein FDR levels. To
properly function, it expects to have access to the following data

```{r msnidcols, echo=FALSE}
sort(MSnID:::.mustBeColumns)
```

which are indeed present in our data:

```{r msnidnames}
names(msnid)
```

Here, we summarise a few steps and redirect the reader to the
package's vignette for more details:


### Analysis of peptide sequences

Cleaning irregular cleavages at the termini of the peptides and
missing cleavage site within the peptide sequences. The following two
function call create the new `numMisCleavages` and `numIrregCleavages`
columns in the `MSnID` object

```{r msnidtermini}
msnid <- assess_termini(msnid, validCleavagePattern="[KR]\\.[^P]")
msnid <- assess_missed_cleavages(msnid, missedCleavagePattern="[KR](?=[^P$])")
```

### Trimming the data

Now, we can use the `apply_filter` function to effectively apply
filters. The strings passed to the function represent expressions that
will be evaluated, thus keeping only PSMs that have 0 irregular
cleavages and 2 or less missed cleavages.

```{r msnidtrim}
msnid <- apply_filter(msnid, "numIrregCleavages == 0")
msnid <- apply_filter(msnid, "numMissCleavages <= 2")
show(msnid)
```

### Parent ion mass errors

Using `"calculatedMassToCharge"` and `"experimentalMassToCharge"`, the
`mass_measurement_error` function calculates the parent ion mass
measurement error in parts per million.

```{r msnidppm1}
summary(mass_measurement_error(msnid))
```

We then filter any matches that do not fit the +/- 20 ppm tolerance

```{r msnidppm2}
msnid <- apply_filter(msnid, "abs(mass_measurement_error(msnid)) < 20")
summary(mass_measurement_error(msnid))
```

### Filtering criteria

Filtering of the identification data will rely on

* -log10 transformed MS-GF+ Spectrum E-value, reflecting the goodness
  of match experimental and theoretical fragmentation patterns

```{r filt1}
msnid$msmsScore <- -log10(msnid$`MS-GF:SpecEValue`)
```

* the absolute mass measurement error (in ppm units) of the parent ion

```{r filt2}
msnid$absParentMassErrorPPM <- abs(mass_measurement_error(msnid))
```

### Setting filters

MS2 filters are handled by a special `MSnIDFilter` class objects, where
individual filters are set by name (that is present in `names(msnid)`)
and comparison operator (>, <, = , ...)  defining if we should retain
hits with higher or lower given the threshold and finally the
threshold value itself.

```{r filt3}
filtObj <- MSnIDFilter(msnid)
filtObj$absParentMassErrorPPM <- list(comparison="<", threshold=10.0)
filtObj$msmsScore <- list(comparison=">", threshold=10.0)
show(filtObj)
```

We can then evaluate the filter on the identification data object,
which return the false discovery rate and number of retained
identifications for the filtering criteria at hand.

```{r filt4}
evaluate_filter(msnid, filtObj)
```

### Filter optimisation

Rather than setting filtering values by hand, as shown above, these
can be set automatically to meet a specific false discovery rate.

```{r optim1}
filtObj.grid <- optimize_filter(filtObj, msnid, fdr.max=0.01,
                                method="Grid", level="peptide",
                                n.iter=500)
show(filtObj.grid)
```

```{r optim2}
evaluate_filter(msnid, filtObj.grid)
```

Filters can eventually be applied (rather than just evaluated) using
the `apply_filter` function.

```{r optim3}
msnid <- apply_filter(msnid, filtObj.grid)
show(msnid)
```

And finally, identifications that matched decoy and contaminant
protein sequences are removed

```{r optim4}
msnid <- apply_filter(msnid, "isDecoy == FALSE")
msnid <- apply_filter(msnid, "!grepl('Contaminant',accession)")
show(msnid)
```

### Export `MSnID` data

The resulting filtered identification data can be exported to a
`data.frame` (or to a dedicated `MSnSet` data structure from the
`MSnbase` package) for quantitative MS data, described below, and
further processed and analyses using appropriate statistical tests.

```{r}
head(psms(msnid))
```
