<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 1 Preamble | R/Bioconductor for Mass Spectrometry and Proteomics" />
<meta property="og:type" content="book" />





<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2021-03-13" />

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="/usr/share/javascript/mathjax/MathJax.js?config=TeX-AMS_CHTML-full" type="text/javascript"></script>

<meta name="description" content="Chapter 1 Preamble | R/Bioconductor for Mass Spectrometry and Proteomics">

<title>Chapter 1 Preamble | R/Bioconductor for Mass Spectrometry and Proteomics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }

code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">R/Bioconductor for Mass Spectrometry and Proteomics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a id="active-page" href="index.html"><span class="toc-section-number">1</span> Preamble</a><ul class="toc-sections">
<li class="toc"><a href="#about-this-course"> About this course</a></li>
<li class="toc"><a href="#the-r-for-mass-spectrometry-initiative"> The R for Mass Spectrometry initiative</a></li>
<li class="toc"><a href="#setup"> Setup</a></li>
</ul>
<a href="sec-msintro.html"><span class="toc-section-number">2</span> Introduction</a>
<a href="raw-ms-data.html"><span class="toc-section-number">3</span> Raw MS data</a>
<a href="identification-data.html"><span class="toc-section-number">4</span> Identification data</a>
<a href="sec-quant.html"><span class="toc-section-number">5</span> Quantitative data</a>
<a href="sec-si.html"><span class="toc-section-number">6</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<!--bookdown:title:end--><!--bookdown:title:start--><html><body><div id="preamble" class="section level1">
<h1>
<span class="header-section-number">Chapter 1</span> Preamble</h1>
<div id="about-this-course" class="section level2">
<h2>
<span class="header-section-number">1.1</span> About this course</h2>
<p>This course will introduce participants to the analysis and
exploration of mass spectrometry (MS) based proteomics data using R
and Bioconductor. The course will cover all levels of MS data, from
raw data to identification and quantitation data, up to the
statistical interpretation of a typical shotgun MS experiment and will
focus on hands-on tutorials. At the end of this course, the
participants will be able to manipulate MS data in R and use existing
packages for their exploratory and statistical proteomics data
analysis.</p>
<div id="targeted-audience-and-assumed-background" class="section level3 unnumbered">
<h3>Targeted audience and assumed background</h3>
<p>The course is targeted to either proteomics practitioners or data
analysts/bioinformaticians that would like to learn how to use R and
Bioconductor to analyse proteomics data. Familiarity with MS or
proteomics in general is desirable, but not essential as we will walk
through and describe a typical MS data as part of learning about the
tools. Participants need to have a working knowledge of R (R syntax,
commonly used functions, basic data structures such as data frames,
vectors, matrices, â€¦ and their manipulation). Familiarity with other
Bioconductor omics data classes and the tidyverse syntax is useful,
but not required.</p>
</div>
<div id="program" class="section level3 unnumbered">
<h3>Program</h3>
<ul>
<li><p>In the first part of this course, we will focus on raw MS data,
including how mass spectrometry works, how raw MS data looks like,
MS data formats, and how to extract, manipulate and visualise raw
data.</p></li>
<li><p>The second part will focus in identification data, how to combine
them with raw data, quantitation of MS data, and introduce data
structure of quantitative proteomics data.</p></li>
<li><p>The last part will focus on quantitative proteomics, including data
structures, data processing, visualisation statistical analysis to
identify differentially expression proteins between two groups.</p></li>
</ul>
<p>The matriel from this course is compiled from various documents, from
the <a href="https://lgatto.github.io/bioc-ms-prot/lab.html"><em>bioc-ms-prot</em></a>
and
<a href="https://lgatto.github.io/bioc-ms-prot/csama2019-lab.html"><em>CSAMA</em></a>
labs. See also <a href="%5Bhttps://github.com/lgatto/2020-02-17-RProt-Berlin%5D">a previous
interation</a> of
this course.</p>
</div>
</div>
<div id="the-r-for-mass-spectrometry-initiative" class="section level2">
<h2>
<span class="header-section-number">1.2</span> The R for Mass Spectrometry initiative</h2>
<p>The <a href="https://www.rformassspectrometry.org/">R for Mass Spectrometry
initiative</a> is a relatively
recent project. Its aim is to provide efficient, thoroughly
documented, tested and flexible R software for the analysis and
interpretation of high throughput mass spectrometry assays, including
proteomics and metabolomics experiments. The project formalises the
longtime collaborative development efforts of its core members under
the <em>R for Mass Spectrometry</em> organisation to facilitate dissemination
and accessibility of their work.</p>
<p>We will be making use of several of these packages in this course.</p>
</div>
<div id="setup" class="section level2">
<h2>
<span class="header-section-number">1.3</span> Setup</h2>
<p>The participants should set up <a href="https://uclouvain-cbio.github.io/WSBIM1207/sec-rrstudio.html">R and
RStudio</a>
and be familiar with <a href="https://uclouvain-cbio.github.io/WSBIM1207/sec-startr.html">R
basics</a></p>
<p>Familiarity with
<a href="https://uclouvain-cbio.github.io/WSBIM1207/sec-bioinfo.html#sec:bioconductor">Bioconductor</a>
is useful, but not necessary. We will be learning about different
types of
<a href="https://uclouvain-cbio.github.io/WSBIM1322/sec-obj.html">objects</a>
related to mass spectrometry and proteomics throughout the course. No
experience in object-oriented programming is necessary.</p>
<p>Package installation instructions:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" title="1"><span class="cf">if</span> (<span class="op">!</span><span class="kw">requireNamespace</span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>))</a>
<a class="sourceLine" id="cb1-2" title="2">    <span class="kw">install.packages</span>(<span class="st">"BiocManager"</span>)</a>
<a class="sourceLine" id="cb1-3" title="3"></a>
<a class="sourceLine" id="cb1-4" title="4">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="dt">version =</span> <span class="st">"3.12"</span>)</a>
<a class="sourceLine" id="cb1-5" title="5"></a>
<a class="sourceLine" id="cb1-6" title="6"><span class="co">###Install the required libraries</span></a>
<a class="sourceLine" id="cb1-7" title="7">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"msdata"</span>)</a>
<a class="sourceLine" id="cb1-8" title="8">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"mzR"</span>)</a>
<a class="sourceLine" id="cb1-9" title="9">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"lgatto/ProtGenerics"</span>)</a>
<a class="sourceLine" id="cb1-10" title="10">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"RforMassSpectrometry/MsCoreUtils"</span>)</a>
<a class="sourceLine" id="cb1-11" title="11">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"RforMassSpectrometry/QFeatures"</span>)</a>
<a class="sourceLine" id="cb1-12" title="12">BiocManager<span class="op">::</span><span class="kw">install</span>(<span class="st">"RforMassSpectrometry/PSM"</span>)</a></code></pre></div>

</div>
</div></body></html>

<p style="text-align: center;">
<a href="sec-msintro.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2021-03-13
</p>
</div>
</div>



</body>
</html>
