<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta name="viewport" content="width=device-width, initial-scale=1" />
<meta property="og:title" content="Chapter 2 Introduction | R/Bioconductor for Mass Spectrometry and Proteomics" />
<meta property="og:type" content="book" />





<meta name="author" content="Laurent Gatto" />

<meta name="date" content="2021-03-14" />


<meta name="description" content="Chapter 2 Introduction | R/Bioconductor for Mass Spectrometry and Proteomics">

<title>Chapter 2 Introduction | R/Bioconductor for Mass Spectrometry and Proteomics</title>

<link href="libs/tufte-css-2015.12.29/tufte.css" rel="stylesheet" />
<link href="libs/tufte-css-2015.12.29/envisioned.css" rel="stylesheet" />
<link href="libs/msmb-css-0/msmb.css" rel="stylesheet" />
<script>
function toggle_visibility(id1, id2) {
var e = document.getElementById(id1);
var f = document.getElementById(id2);

e.style.display = ((e.style.display!='none') ? 'none' : 'block');

if(f.classList.contains('fa-plus-square')) {
    f.classList.add('fa-minus-square')
    f.classList.remove('fa-plus-square')
} else {
    f.classList.add('fa-plus-square')
    f.classList.remove('fa-minus-square')
}

}
</script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }

code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(title);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  { background-color: #f8f8f8; }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />

</head>

<body>



<div class="row">
<div class="col-sm-12">
<div id="TOC">
<ul class="navbar">
<li class="msmb"><p class="title">R/Bioconductor for Mass Spectrometry and Proteomics<p><p class="author">Laurent Gatto</p>
<li class="dropdown" style="float:right">
<a href="javascript:void(0)" class="dropbtn">&#x25BE; Chapters</a>
<div class="dropdown-content">
<a href="index.html"><span class="toc-section-number">1</span> Preamble</a>
<a id="active-page" href="sec-msintro.html"><span class="toc-section-number">2</span> Introduction</a><ul class="toc-sections">
<li class="toc"><a href="#how-does-mass-spectrometry-work"> How does mass spectrometry work?</a></li>
<li class="toc"><a href="#accessing-data"> Accessing data</a></li>
</ul>
<a href="raw-ms-data.html"><span class="toc-section-number">3</span> Raw MS data</a>
<a href="identification-data.html"><span class="toc-section-number">4</span> Identification data</a>
<a href="sec-quant.html"><span class="toc-section-number">5</span> Quantitative data</a>
<a href="sec-si.html"><span class="toc-section-number">6</span> Session information</a>
</div>
</li>
</ul>
</div>
</div>
</div>
<div class="row">
<div class="col-sm-12">
<!DOCTYPE html PUBLIC "-//W3C//DTD HTML 4.0 Transitional//EN" "http://www.w3.org/TR/REC-html40/loose.dtd">
<html><body><div id="sec:msintro" class="section level1">
<h1>
<span class="header-section-number">Chapter 2</span> Introduction</h1>
<div id="how-does-mass-spectrometry-work" class="section level2">
<h2>
<span class="header-section-number">2.1</span> How does mass spectrometry work?</h2>
<p>Mass spectrometry (MS) is a technology that <em>separates</em> charged
molecules (ions) based on their mass to charge ratio (M/Z). It is
often coupled to chromatography (liquid LC, but can also be gas-based
GC). The time an analytes takes to elute from the chromatography
column is the <em>retention time</em>.</p>
<div class="figure">
<img src="img/chromatogram.png" alt="A chromatogram, illustrating the total amount of analytes over the retention time."><p class="caption">A chromatogram, illustrating the total amount of analytes over the retention time.</p>
</div>
<p>An mass spectrometer is composed of three components:</p>
<ol style="list-style-type: decimal">
<li>The <em>source</em>, that ionises the molecules: examples are Matrix-assisted
laser desorption/ionisation (MALDI) or electrospray ionisation.
(ESI)</li>
<li>The <em>analyser</em>, that separates the ions: Time of flight (TOF) or Orbitrap.</li>
<li>The <em>detector</em> that quantifies the ions.</li>
</ol>
<p>When using mass spectrometry for proteomics, the proteins are first
digested with a protease such as trypsin. In mass shotgun proteomics,
the analytes assayed in the mass spectrometer are peptides.</p>
<p>Often, ions are subjected to more than a single MS round. After a
first round of separation, the peaks in the spectra, called MS1
spectra, represent peptides. At this stage, the only information we
possess about these peptides are their retention time and their
mass-to-charge (we can also infer their charge be inspecting their
isotopic envelope, i.e the peaks of the individual isotopes, see
below), which is not enough to infer their identify (i.e. their
sequence).</p>
<p>In MSMS (or MS2), the settings of the mass spectrometer are set
automatically to select a certain number of MS1 peaks (for example
20). Once a narrow M/Z range has been selected (corresponding to one
high-intensity peak, a peptide, and some background noise), it is
fragmented (using for example collision-induced dissociation (CID),
higher energy collisional dissociation (HCD) or electron-transfer
dissociation (ETD)). The fragment ions are then themselves separated
in the analyser to produce a MS2 spectrum. The unique fragment ion
pattern can then be used to infer the peptide sequence using de novo
sequencing (when the spectrum is of high enough quality) of using a
search engine such as, for example Mascot, MSGF+, …, that will match
the observed, experimental spectrum to theoratical spectra (see
details below).</p>
<div class="figure" style="text-align: center">
<span id="fig:unnamed-chunk-3"></span>
<p class="caption marginnote shownote">
Figure 2.1: Schematics of a mass spectrometer and two rounds of MS.
</p>
<img src="img/SchematicMS2.png" alt="Schematics of a mass spectrometer and two rounds of MS." width="100%">
</div>
<p>The animation below show how 25 ions different ions (i.e. having
different M/Z values) are separated throughout the MS analysis and are
eventually detected (i.e. quantified). The final frame shows the
hypothetical spectrum.</p>
<div class="figure" style="text-align: center">
<span id="fig:unnamed-chunk-4"></span>
<p class="caption marginnote shownote">
Figure 2.2: Separation and detection of ions in a mass spectrometer.
</p>
<img src="img/mstut.gif" alt="Separation and detection of ions in a mass spectrometer." width="100%">
</div>
<p>The figures below illustrate the two rounds of MS. The spectrum on the
left is an MS1 spectrum acquired after 21 minutes and 3 seconds of
elution. 10 peaks, highlited by dotted vertical lines, were selected
for MS2 analysis. The peak at M/Z 460.79 (488.8) is highlighted by a
red (orange) vertical line on the MS1 spectrum and the fragment
spectra are shown on the MS2 spectrum on the top (bottom) right
figure.</p>
<div class="figure" style="text-align: center">
<span id="fig:unnamed-chunk-5"></span>
<p class="caption marginnote shownote">
Figure 2.3: Parent ions in the MS1 spectrum (left) and two sected fragment ions MS2 spectra (right)
</p>
<img src="img/MS1-MS2-spectra.png" alt="Parent ions in the MS1 spectrum (left) and two sected fragment ions MS2 spectra (right)" width="100%">
</div>
<p>The figures below represent the 3 dimensions of MS data: a set of
spectra (M/Z and intensity) of retention time, as well as the
interleaved nature of MS1 and MS2 (and there could be more levels)
data.</p>
<div class="figure" style="text-align: center">
<span id="fig:unnamed-chunk-6"></span>
<p class="caption marginnote shownote">
Figure 2.4: MS1 spectra over retention time.
</p>
<img src="img/F02-3D-MS1-scans-400-1200-lattice.png" alt="MS1 spectra over retention time." width="100%">
</div>
<div class="figure" style="text-align: center">
<span id="fig:unnamed-chunk-7"></span>
<p class="caption marginnote shownote">
Figure 2.5: MS2 spectra interleaved between two MS1 spectra..
</p>
<img src="img/F02-3D-MS1-MS2-scans-100-1200-lattice.png" alt="MS2 spectra interleaved between two MS1 spectra.." width="100%">
</div>
</div>
<div id="accessing-data" class="section level2">
<h2>
<span class="header-section-number">2.2</span> Accessing data</h2>
<div id="from-the-proteomexchange-database" class="section level3 unnumbered">
<h3>From the ProteomeXchange database</h3>
<p>MS-based proteomics data is disseminated through the
<a href="http://www.proteomexchange.org/">ProteomeXchange</a> infrastructure,
which centrally coordinates submission, storage and dissemination
through multiple data repositories, such as the
<a href="https://www.ebi.ac.uk/pride/archive/">PRoteomics IDEntifications (PRIDE)</a>
database at the EBI for mass spectrometry-based experiments (including
quantitative data, as opposed as the name suggests),
<a href="http://www.peptideatlas.org/passel/">PASSEL</a> at the ISB for Selected
Reaction Monitoring (SRM, i.e. targeted) data and the
<a href="http://massive.ucsd.edu/ProteoSAFe/static/massive.jsp">MassIVE</a>
resource. These data can be downloaded within R using the
<em><a href="https://bioconductor.org/packages/3.12/rpx">rpx</a></em> package.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" title="1"><span class="kw">library</span>(<span class="st">"rpx"</span>)</a>
<a class="sourceLine" id="cb2-2" title="2"><span class="kw">pxannounced</span>()</a></code></pre></div>
<pre><code>## 15 new ProteomeXchange annoucements</code></pre>
<pre><code>##     Data.Set Publication.Data Message
## 1  PXD017953             &lt;NA&gt;     New
## 2  PXD020016             &lt;NA&gt;     New
## 3  PXD020399             &lt;NA&gt;     New
## 4  PXD020714             &lt;NA&gt;     New
## 5  PXD018932             &lt;NA&gt;     New
## 6  PXD023139             &lt;NA&gt;     New
## 7  PXD024706             &lt;NA&gt;     New
## 8  PXD024705             &lt;NA&gt;     New
## 9  PXD013791             &lt;NA&gt;     New
## 10 PXD022739             &lt;NA&gt;     New
## 11 PXD020167             &lt;NA&gt;     New
## 12 PXD011568             &lt;NA&gt;     New
## 13 PXD020167             &lt;NA&gt;     New
## 14 PXD024381             &lt;NA&gt;     New
## 15 PXD024674             &lt;NA&gt;     New</code></pre>
<p>Using the unique <code>PXD000001</code> identifier, we can retrieve the relevant
metadata that will be stored in a <code>PXDataset</code> object. The names of the
files available in this data can be retrieved with the <code>pxfiles</code>
accessor function.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" title="1">px &lt;-<span class="st"> </span><span class="kw">PXDataset</span>(<span class="st">"PXD000001"</span>)</a>
<a class="sourceLine" id="cb5-2" title="2">px</a></code></pre></div>
<pre><code>## Object of class "PXDataset"
##  Id: PXD000001 with 11 files
##  [1] 'F063721.dat' ... [11] 'erwinia_carotovora.fasta'
##  Use 'pxfiles(.)' to see all files.</code></pre>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" title="1"><span class="kw">pxfiles</span>(px)</a></code></pre></div>
<pre><code>##  [1] "F063721.dat"                                                         
##  [2] "F063721.dat-mztab.txt"                                               
##  [3] "PRIDE_Exp_Complete_Ac_22134.xml.gz"                                  
##  [4] "PRIDE_Exp_mzData_Ac_22134.xml.gz"                                    
##  [5] "PXD000001_mztab.txt"                                                 
##  [6] "README.txt"                                                          
##  [7] "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML" 
##  [8] "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzXML"
##  [9] "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01.mzXML"         
## [10] "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01.raw"           
## [11] "erwinia_carotovora.fasta"</code></pre>
<p>Other metadata for the <code>px</code> data set:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" title="1"><span class="kw">pxtax</span>(px)</a></code></pre></div>
<pre><code>## [1] "Erwinia carotovora"</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" title="1"><span class="kw">pxurl</span>(px)</a></code></pre></div>
<pre><code>## [1] "ftp://ftp.pride.ebi.ac.uk/pride/data/archive/2012/03/PXD000001"</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" title="1"><span class="kw">pxref</span>(px)</a></code></pre></div>
<pre><code>## [1] "Gatto L, Christoforou A. Using R and Bioconductor for proteomics data analysis. Biochim Biophys Acta. 2013 May 18. doi:pii: S1570-9639(13)00186-6. 10.1016/j.bbapap.2013.04.032"</code></pre>
<p>Data files can then be downloaded with the <code>pxget</code> function. Below, we
retrieve the raw data file. The file is
downloaded<label for="tufte-sn-1" class="margin-toggle sidenote-number">1</label><input type="checkbox" id="tufte-sn-1" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">1</span> If the file is already available, it is not downloaded a second time.</span>
in the working directory and the name of the file is return by the
function and stored in the <code>mzf</code> variable for later use.<label for="tufte-sn-2" class="margin-toggle sidenote-number">2</label><input type="checkbox" id="tufte-sn-2" class="margin-toggle"><span class="sidenote"><span class="sidenote-number">2</span> This and other files are also availabel in the <code>msdata</code> package, described below</span></p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" title="1">fn &lt;-<span class="st"> "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML"</span></a>
<a class="sourceLine" id="cb15-2" title="2">mzf &lt;-<span class="st"> </span><span class="kw">pxget</span>(px, fn)</a></code></pre></div>
<pre><code>## Loading TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML from cache.</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" title="1">mzf</a></code></pre></div>
<pre><code>## [1] "~/.cache/rpx/10c1f4e4c6980_TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML"</code></pre>
</div>
<div id="data-packages" class="section level3 unnumbered">
<h3>Data packages</h3>
<p>Some data are also distributed through dedicated packages. The
<em><a href="https://bioconductor.org/packages/3.12/msdata">msdata</a></em>, for example, provides some
general raw data files relevant for both proteomics and
metabolomics.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" title="1"><span class="kw">library</span>(<span class="st">"msdata"</span>)</a>
<a class="sourceLine" id="cb19-2" title="2"><span class="co">## proteomics raw data</span></a>
<a class="sourceLine" id="cb19-3" title="3"><span class="kw">proteomics</span>()</a></code></pre></div>
<pre><code>## [1] "MRM-standmix-5.mzML.gz"                                                
## [2] "MS3TMT10_01022016_32917-33481.mzML.gz"                                 
## [3] "MS3TMT11.mzML"                                                         
## [4] "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzML.gz"
## [5] "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01.mzML.gz"</code></pre>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" title="1"><span class="co">## proteomics identification data</span></a>
<a class="sourceLine" id="cb21-2" title="2"><span class="kw">ident</span>()</a></code></pre></div>
<pre><code>## [1] "TMT_Erwinia_1uLSike_Top10HCD_isol2_45stepped_60min_01-20141210.mzid"</code></pre>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb23-1" title="1"><span class="co">## quantitative data</span></a>
<a class="sourceLine" id="cb23-2" title="2"><span class="kw">quant</span>()</a></code></pre></div>
<pre><code>## [1] "cptac_a_b_peptides.txt"</code></pre>
<p>More often, such <em>experiment packages</em> distribute processed data; an
example of such is the <em><a href="https://bioconductor.org/packages/3.12/pRolocdata">pRolocdata</a></em>
package, that offers quantitative proteomics data.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" title="1">pRolocdata<span class="op">::</span><span class="kw">pRolocdata</span>()</a></code></pre></div>

</div>
</div>
</div></body></html>

<p style="text-align: center;">
<a href="index.html"><button class="btn btn-default">Previous</button></a>
<a href="raw-ms-data.html"><button class="btn btn-default">Next</button></a>
</p>
<p class="build-date">Page built: 
2021-03-14
</p>
</div>
</div>



</body>
</html>
